### DISCUSSION TITLE
How would I move the ext4.vhdx from my C drive to a different drive?

### DISCUSSION THREAD

**Discussion by xela-nosretap** [microsoft/WSL#8637] [Q&A] [answered]
My C drive is getting a bit full, I'd like to move the nearly 40 GB that my Debian distro is taking up on my C drive and move it to a different drive with a bit more space on it.

**Comment by NotTheDr01ds** [comment, marked as answer]
From PowerShell (normal user, *not* Administrator):

```PowerShell
# Modify variables below as needed/desired
$WSL_ROOT="D:\WSL"
$WSL_IMAGE_PATH="${WSL_ROOT}\images"
$WSL_OLD_DISTRO_NAME="Debian"
$WSL_NEW_DISTRO_NAME="Debian_WSL2"
$WSL_INSTANCE_PATH="${WSL_ROOT}\instances\${WSL_NEW_DISTRO_NAME"

mkdir $WSL_IMAGE_PATH
mkdir $WSL_INSTANCE_PATH

# Create a backup/export of the existing Debian in `D:\WSL\images\Debian.backup.tar`
wsl --shutdown
wsl --export $WSL_OLD_DISTRO_NAME "${WSL_IMAGE_PATH}\${WSL_OLD_DISTRO_NAME}.backup.tar"

# Import that backup into the new location/distribution:
wsl --import $WSL_NEW_DISTRO_NAME $WSL_INSTANCE_PATH "${WSL_IMAGE_PATH}\${WSL_OLD_DISTRO_NAME}.backup.tar" --version 2

# Start the new distro:
wsl ~ -d $WSL_NEW_DISTRO_NAME
```

From inside the new Debian distro:

```bash
sudo -e /etc/wsl.conf
# Yes, you are already root, but `sudo -e` is just a fairly safe way to invoke the default editor regardless of distribution
```

Add the following:

```ini
[user]
default=<your_normal_username>
```

 Save and exit the editor

`exit` the shell/WSL.

Back in the same PowerShell session (so all the variables are still intact):

```
wsl --terminate $WSL_NEW_DISTRO_NAME
wsl --set-default $WSL_NEW_DISTRO_NAME
wsl ~
```

You should now be running as your normal user in the new Debian location.

To confirm, `wsl.exe -l -v` should show that the new distro is running, and the old one is not.  Additionally, from PowerShell:

```
Get-ChildItem HKCU:\Software\Microsoft\Windows\CurrentVersion\Lxss\ | ForEach-Object { (Get-ItemProperty $_.PSPATH) | Select-Object DistributionName,BasePath }
```

... will show the installed distributions and their locations.

After confirming that everything is working properly:

**Warning**: This is a destructive and unrecoverable operation

```
$WSL_OLD_DISTRO_NAME="Debian"
wsl --unregister $WSL_OLD_DISTRO_NAME
```

You can also uninstall the Debian app from the Store as well, but I personally recommend keeping it around.  Having it still installed means that you can quickly spin up a new Debian to "test something out" with:

```
debian.exe
```

Set the username/password, `sudo apt update && sudo apt upgrade -y`, and you have a fresh, clean installation that you can install something in to test, then:

```
wsl --unregister Debian
```

to throw it away.

**Reply by lars18th** [reply]
Exporting and Importing is _not_ a "MOVE" action. 👎

**Comment by lars18th** [comment]
Hi,

We know how to _export_ the distribution and how to then move it to another drive and registrer it again. However, this method is error prone. Why? Because a lot of circunstances can fail in the exporting procedure. Futhermore, the registering again can produce some troubles.

So, why not provide a simple and direct method to a simple _move_ command? Like `wsl --move-to E:\wsl-store\ Debian`.
This process will be a simple copy of the "debian.vhdx" file to the destination folder. Plus the changes on the registry (aka `HKCU\Software\Microsoft\Windows\CurrentVersion\Lxss\`). If this will be native supported by the official tool, then the users could store the files in any location without effort.

So, please consider it.

**Comment by go-xoxo** [comment]
It's as simple as shutting down the WSL distro, moving the ext4.vhdx file, and changing the BasePath to the new directory. There's no need to export, import, or unregister. However, a --move command that performs these steps would be a nice addition. Fiddling with the registry is always a bit delicate.

**Comment by lars18th** [comment]
Hi @go-xoxo ,

> It's as simple as shutting down the WSL distro, moving the ext4.vhdx file, and changing the BasePath to the new directory. There's no need to export, import, or unregister. However, a --move command that performs these steps would be a nice addition. Fiddling with the registry is always a bit delicate.

The "BasePath" is stored in the Registry. So the _only_ method to move the ext4.vhdx file is editing the info inside the registry. Therefore, I hope the developers will add the method `wsl --move-to BasePath distribution`. This is a must-have! 🤔
